controller/gateway/controller_reconciler_utils.go

```diff
@@ -555,11 +555,18 @@ func supportedRoutesByProtocol() map[gatewayv1.ProtocolType]map[gatewayv1.Kind]s
 // It also sets the listeners Programmed condition by setting the underlying
 // Listener Programmed status to false.
 func (g *gatewayConditionsAndListenersAwareT) initProgrammedAndListenersStatus() {
-	k8sutils.InitProgrammed(g)
+	k8sutils.SetCondition(
```

Comment:
I can see that this change pretty much only sets the generation in addition to what has been done previously but what I'm concerned about is that we seem to always set this condition regardless if that condition has already been set on `Gateway` or not.

https://github.com/Kong/gateway-operator/blob/dcb56adfdf81438a191b4a1533f565a08c6fb704/controller/gateway/controller.go#L169-L186

Shouldn't that be checking if `Programmed` has been set and only initialize it if it's not there yet? Just like we do it for listeners in https://github.com/Kong/gateway-operator/blob/508daaf373bf6ff7b9b06b81b9ded9e976516d28/controller/gateway/controller_reconciler_utils.go#L560-L562 ?

Reply:
It needs to be like this, as the programmed condition is initialized as false and set to true only in case all the other conditions are true. This means that when you have a programmed Gateway and the ControlPlane gets deleted, the programmed condition is set to false again thanks to this initialization a priori. Once the ControlPlane will be provisioned and ready again, the programmed condition will transition to true, as the Gateway Controller will get to the bottom of the reconciliation loop. 

Does it make sense to you?